### PR TITLE
refactor(builder): large refactor of bundle sender state machine and metrics

### PR DESCRIPTION
**Pull request by dancoombs**
## Proposed Changes

  - Refactor state machine
  - Refactor metrics by saving an object with common components

### CODE REVIEW

File: crates/builder/src/bundle_sender.rs [hunk 1]
```diff
@@ -57,10 +57,11 @@ pub(crate) struct BundleSenderImpl<UO, P, E, T, C> {
     beneficiary: Address,
     proposer: P,
     entry_point: E,
-    transaction_tracker: T,
+    transaction_tracker: Option<T>,
```
Comment (0xfourzerofour) [inline crates/builder/src/bundle_sender.rs:60]: when would this be None?

File: crates/builder/src/bundle_sender.rs [hunk 2]
```diff
@@ -464,48 +174,266 @@ where
...
+                TrackerUpdate::Mined {
+                    block_number,
+                    attempt_number,
+                    gas_limit,
+                    gas_used,
+                    tx_hash,
+                    nonce,
+                    ..
+                } => {
+                    // mined!
```
Comment (0xfourzerofour) [inline crates/builder/src/bundle_sender.rs:289]: nit: remove comment

File: crates/builder/src/bundle_sender.rs [hunk 3]
```diff
@@ -464,48 +174,266 @@ where
...
+
+        Ok(())
+    }
+
+    async fn handle_cancelling_state(
+        &mut self,
+        state: &mut SenderMachineState<T>,
+        inner: CancellingState,
+    ) -> anyhow::Result<()> {
+        // cancel the transaction
```
Comment (0xfourzerofour) [inline crates/builder/src/bundle_sender.rs:347]: Also no sure if this is needed with the info message or put the comment above the function

File: crates/builder/src/bundle_sender.rs [hunk 4]
```diff
@@ -124,313 +122,25 @@ where
...
-                                fee_increase_count + 1
-                            );
-                            state = State::Cancelling(fee_increase_count + 1);
-                        }
-                    }
-                }
-            }
+        // initial state
+        let mut state =
+            SenderMachineState::new(sender_trigger, self.transaction_tracker.take().unwrap());
```
Comment (0xfourzerofour) [inline crates/builder/src/bundle_sender.rs:135]: If this is always going to be Some, what is the ideology for making it optional?

File: crates/builder/src/bundle_sender.rs [hunk 5]
```diff
@@ -124,313 +122,25 @@ where
...
 
-            // send result to manual caller
-            if let Some(res) = send_bundle_result.take() {
-                if let Some(r) = send_bundle_response.take() {
-                    if r.send(res).is_err() {
-                        error!("Failed to send bundle result to manual caller");
-                    }
+        loop {
+            match self.step_state(&mut state).await {
```
Comment (0xfourzerofour) [inline crates/builder/src/bundle_sender.rs:138]: if we are not doing anything with the Ok value, we could probably use an `is_err` of a `let if Err(e)` to remove unused lines

**Review by 0xfourzerofour — APPROVED** [review]
Looks good just a few comments

File: crates/builder/src/bundle_sender.rs (hunk 1 above)
Comment (dancoombs) [reply crates/builder/src/bundle_sender.rs:60]: The transaction tracker isn't `Clone`, and I want to be able to move it into the `SenderMachineState` when `send_bundles_in_loop` is called, as its easier to track there. This was the easiest way to do that without making any of the pub `new` functions async.

File: crates/builder/src/bundle_sender.rs (hunk 4 above)
Comment (dancoombs) [reply crates/builder/src/bundle_sender.rs:135]: So that we can move out of the optional, the optional is left with `None` after this is called.

**Pull request merged**